### DISCUSSION TITLE
How to write a Dockerfile correctly

### DISCUSSION THREAD

**Discussion by robml** [pdm-project/pdm#896] [Q&A]
Just playing around with `pdm` given how little it is mentioned anywhere else, wanted to ask if this looks about right for a Dockerfile:
```
# syntax=docker/dockerfile:1
FROM python:3
ENV PYTHONDONTWRITEBYTECODE=1
ENV PYTHONUNBUFFERED=1
WORKDIR /code
COPY pyproject.toml pdm.lock /code/
RUN pip install pdm && pip add --prod --no-lock --no-editable
COPY . /code/
```
I adopted this from [the Docker website](https://docs.docker.com/samples/django/) and changed it from a regular `pip` and `requirements.txt` install to the one seen on the official `pdm` website for a multistage Docker set up. Any ideas?

**Comment by pawamoy** [comment]
I believe you meant `&& pdm install` instead of `&& pip add`?

**Reply by pawamoy** [reply]
Also, I copy the code onto the image before I run `pdm install` so that my project is installed as well in the `__pypackages__` folder, which I finally copy into a final image (multi-stage build) where I add it to PYTHONPATH. Basically this https://pdm.fming.dev/usage/advanced/#use-pdm-in-a-multi-stage-dockerfile

**Reply by robml** [reply]
That's right, I corrected it in a later version. I might do that to be honest. I don't know too much about Docker builds I am still getting the hang of it, but when I ran the above (corrected for `pdm install`), I would still end up with an error. Basically I was working on a Django project trying to create some boilerplate, and the only libraries I "added" with `pdm` were Django and the `requests` library. Django which comes with both a database and a web service, was built into two respective containers, however only the database one worked. Inspecting the logs I found `ImportError: Couldn't import Django. Are you sure it's installed and available on your PYTHONPATH environment variable? Did you forget to activate a virtual environment?` this error message leading me to wonder if there was a problem in my Dockerfile originally. Is the multistage the only way to build a docker file with PDM?

**Reply by pawamoy** [reply]
You don't need multi-stage, no. It's just best-practice: the less you have in the final image, the lighter and more secure it is.

**Reply by robml** [reply]
How would I resolve the above in that case then? In a pip setting I simply would add to the `requirements.txt` file and let the container build the necessary packages. In the above case simple substitution doesn't seem to be working, does `pdm` have a different way of working with `PYTHONPATH` than `pip`, and how should we then address that?

**Reply by pawamoy** [reply]
Are you trying to use *both* pip and PDM as package managers at the same time? I'd not recommend it if you're not comfortable with PDM/Docker as it surely complicates things.

With pip, you'd typically:

1. install deps in a virtualenv, and keep it activated in the RUN or ENTRYPOINT instructions
2. or install deps in the system site-packages

While with PDM you'd typically:

1. install deps with `pdm install --prod --no-lock --no-editable`
2. 
    i. use `pdm run python ...` instead of `python` in your RUN or ENTRYPOINT instructions
    ii. or add the `__pypackages__/{major.minor}/lib` folder to PYTHONPATH so `python` can find packages

Does it make more sense?

**Reply by robml** [reply]
Kind of, I am doing active research/reading the docs to understand more throughout this conversation. My intention was to only use `pdm`, the `pip` there is only to install `pdm` originally which I think I've done correctly. Based on what you wrote, I think I have installed the deps correctly, however I don't follow point 2 you make above, considering in my code I didn't have any RUN or ENTRYPOINT instructions that did not use `pdm` or I guess `pip` to install `pdm` (should I have?). I think that second piece of information 2ii is key, I am not sure how I would add it to the PYTHONPATH but I think that would resolve my concern above (any tips?). I guess a follow up is if I do add `__pypackages__/{major.minor}/lib` to the PYTHONPATH while building the container, would that pose any issues when deploying to production or not necessarily since I am simply running all of the code inside an isolated container anyways? Thanks for the quality responses btw.

**Reply by pawamoy** [reply]
Setting PYTHONPATH to `__pypackages__/{major.minor}/lib` should pose no issues at all, no, but of course that highly depends on what you do in the container so I can't guarantee anything.

What you should make sure to understand is: how Python finds modules, and how PDM plays with that. When you run `pdm run ...`, PDM makes sure the PATH and PYTHONPATH variables are updated so you can find packages' scripts on the command line, and packages themselves in the Python interpreter. Another way to find packages (but not their scripts) without `pdm run ...` is to [enable PEP 582 support in your shell](https://pdm.fming.dev/#enable-pep-582-globally). A third way is to manually set PYTHONPATH (and possibly PATH but it's not needed usually in a Docker context) yourself :slightly_smiling_face:

**Reply by robml** [reply]
@pawamoy btw thank you so much for your patience. I came back with fresh eyes and tried it again from scratch and it almost seemed to work. I realize that the problem comes with trying to import the django module (everything else seems to work fine).
I got this traceback from one of the containers I tried to make
```
Traceback (most recent call last):

  File "/code/manage.py", line 11, in main

    from django.core.management import execute_from_command_line

ModuleNotFoundError: No module named 'django'


The above exception was the direct cause of the following exception:


Traceback (most recent call last):

  File "/code/manage.py", line 22, in <module>

    main()

  File "/code/manage.py", line 13, in main

    raise ImportError(

ImportError: Couldn't import Django. Are you sure it's installed and available on your PYTHONPATH environment variable? Did you forget to activate a virtual environment?
```
Note I ran it without setting the PYTHONPATH because I wasn't executing any RUN commands with `python` in my Dockerfile. Do you think that's the cause of the issue here? And if there isn't a PDM specific way to solve the above, is there a resource you recommend I check out or is the Docker docs the best place to try and debug? Thanks again.

**Reply by robml** [reply]
@pawamoy an update on the above for future readers, it was definitely the PYTHONPATH, turns out if you are not running a Multi-Stage build you need to declare it. A weird thing I noticed is that by convention, you'd have the line:
`ENV PYTHONPATH "${PYTHONPATH}:/__pypackages__/{major.minor}/lib"`
However appending to the existing `PYTHONPATH` as above did not seem to work. What did work however almost flawlessly was instead overwriting the `PYTHONPATH` variable in something like this:
`ENV PYTHONPATH __pypackages__/{major.minor}/lib`
For those trying to just do small Docker builds I think outlining this in the docs could save beginners at least a headache @frostming .

Again thanks for the help guys.

**Reply by pawamoy** [reply]
That's what I'm trying to tell you from the beginning: whatever method you choose to run Python, PYTHONPATH *must* be updated somehow. It can be *implicit* (when you use `pdm run ...`) or *explicit* (setting it with `ENV PYTHONPATH=...`). Single stage or multi-stage builds are not relevant here. Sorry for not being clear enough 😅 
Maybe appending didn't work because the variable is empty and you end up with `PYTHONPATH=:/path/to/lib` which might not be valid (starting colon).

**Reply by robml** [reply]
@pawamoy thank you :P I was just confused I think because I haven't worked with Docker much and am trying to change that. As for the empty variable, how would you solve something like that in the future btw?

**Reply by pawamoy** [reply]
Within an `ENV` instruction, I'm not sure 😕
